network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/subscription/settings/NetworkProtectionSettingsStateImpl.kt

```diff
@@ -0,0 +1,72 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.networkprotection.impl.subscription.settings
+
+import com.duckduckgo.common.utils.DispatcherProvider
+import com.duckduckgo.di.scopes.AppScope
+import com.duckduckgo.networkprotection.api.NetworkProtectionState
+import com.duckduckgo.networkprotection.impl.subscription.settings.NetworkProtectionSettingsState.NetPSettingsState
+import com.duckduckgo.networkprotection.impl.subscription.settings.NetworkProtectionSettingsState.NetPSettingsState.Hidden
+import com.duckduckgo.networkprotection.impl.subscription.settings.NetworkProtectionSettingsState.NetPSettingsState.Visible.Activating
+import com.duckduckgo.networkprotection.impl.subscription.settings.NetworkProtectionSettingsState.NetPSettingsState.Visible.Expired
+import com.duckduckgo.networkprotection.impl.subscription.settings.NetworkProtectionSettingsState.NetPSettingsState.Visible.Subscribed
+import com.duckduckgo.networkprotection.impl.subscription.NetpSubscriptionManager
+import com.duckduckgo.networkprotection.impl.subscription.NetpSubscriptionManager.VpnStatus
+import com.duckduckgo.networkprotection.impl.subscription.NetpSubscriptionManager.VpnStatus.ACTIVE
+import com.duckduckgo.networkprotection.impl.subscription.NetpSubscriptionManager.VpnStatus.EXPIRED
+import com.duckduckgo.networkprotection.impl.subscription.NetpSubscriptionManager.VpnStatus.INACTIVE
+import com.duckduckgo.networkprotection.impl.subscription.NetpSubscriptionManager.VpnStatus.INELIGIBLE
+import com.duckduckgo.networkprotection.impl.subscription.NetpSubscriptionManager.VpnStatus.SIGNED_OUT
+import com.duckduckgo.networkprotection.impl.subscription.NetpSubscriptionManager.VpnStatus.WAITING
+import com.duckduckgo.networkprotection.impl.subscription.isActive
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+
+@ContributesBinding(AppScope::class)
+class NetworkProtectionSettingsStateImpl @Inject constructor(
+    private val dispatcherProvider: DispatcherProvider,
+    private val networkProtectionState: NetworkProtectionState,
+    private val netpSubscriptionManager: NetpSubscriptionManager,
+) : NetworkProtectionSettingsState {
+
+    override suspend fun getStateFlow(): Flow<NetPSettingsState> =
+        netpSubscriptionManager.vpnStatus().map { status ->
+            if (!status.isActive()) {
+                // if entitlement check succeeded and not an active subscription then reset state
+                handleRevokedVPNState()
+            }
```

Comment:
Should this logic be here?
If I understand correctly, a user can have an inactive subscription but continue using the VPN as long as they don't open settings? That seems odd/wrong?

Reply:
I'm taking the logic that already exists in `NetworkProtectionAccessStateImpl`  and adding it here.

```
   override suspend fun getStateFlow(): Flow<NetPAccessState> = withContext(dispatcherProvider.io()) {
        netpSubscriptionManager.vpnStatus().map { status ->
            if (!status.isActive()) {
                // if entitlement check succeeded and no entitlement, reset state and hide access.
                handleRevokedVPNState()
                Locked
            } else {
                UnLocked
            }
        }
    }
```

This is only ever called by the current NetP Settings item: `ProSettingNetPViewModel` (`LegacyProSettingNetPViewModel` in this PR)

Once the new Settings is rolled out, `NetworkProtectionAccessStateImpl.getStateFlow` will never be called, so I wanted to ensure the behaviour stays the same when that eventually happens.

Reply:
As discussed on Zoom, this is existing behaviour in the Legacy `ProSettingNetPViewModel` and therefore I'm going to leave it as is so as not to affect existing behaviour.